--- a/src/vmod_header.c
+++ b/src/vmod_header.c
@@ -173,20 +173,18 @@ VCL_VOID
 vmod_append(VRT_CTX, VCL_HEADER hdr, VCL_STRANDS s)
 {
 	struct http *hp;
-	struct strands *st;
+	struct strands st[1];
+	const char *p[s->n + 2];
 	const char *b;

 	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);

 	/* prefix the strand with $hdr_name + space */
-	st = VRT_AllocStrandsWS(ctx->ws, s->n + 2);
-	if (!st) {
-		VRT_fail(ctx, "vmod_head: workspace allocation failure");
-		return;
-	}
-	st->p[0] = hdr->what + 1;
-	st->p[1] = " ";
-	AN(memcpy(st->p + 2, s->p, s->n * sizeof *s->p));
+	p[0] = hdr->what + 1;
+	p[1] = " ";
+	AN(memcpy(p + 2, s->p, s->n * sizeof *s->p));
+	st->n = s->n + 2;
+	st->p = p;

 	b = VRT_StrandsWS(ctx->ws, NULL, st);

--
2.11.0
